### PR TITLE
Switch layer kernel implementation in the config

### PR DESCRIPTION
**Pull request by cathalobrien**
# Describe your changes

This PR makes it possible to switch the implementation of Linear and LayerNorm kernels in the config. 

At the moment we use torch.NN implementation for many layers in Anemoi model e.g. torch.nn.layerNorm, torch.NN.linear. This has the advantage of being available out of the box with torch and portable to many systems (CPU, AMD and Nvidia GPUs). However, other layer implementations might be more efficient for certain hardware, or use different algorithms we want to explore (RMSNorm is an alternate implementation of LayerNorm for instance). 
These might only run on certain systems (e.g. Nvidias transformer_engine.pytorch provides layer implementations optimized for their GPUs). 
Therefore, we'd like to be able to flexibly take advantage of these faster kernels when they're available without losing the ability to fall back to torch.nn for resiliency. 

This PR adds the following block to config/model/<model>.yaml:
```yaml
  layer_kernels:
    LayerNorm:
      #_target_: "transformer_engine.pytorch.LayerNorm"
      _target_: "liger_kernel.transformers.rms_norm.LigerRMSNorm"
      #_target_: "torch.nn.LayerNorm" #the default PyTorch implementation
      _partial_: True
      #Any arguments to your chosen function go here e.g.
      #bias: False
    Linear:
      #_target_: "transformer_engine.pytorch.Linear"
      _target_: "torch.nn.Linear"
      _partial_: True
```
In the future, this syntax could be extended to replace other layers (e.g. mlp) if required. 

The calls to `torch.nn` are then replaced with
```python
- self.layer_norm1 = nn.LayerNorm(num_channels)
+ LayerNorm=layer_kernels['LayerNorm']
+ self.layer_norm1 = LayerNorm(num_channels)
```

You can pass any parameters to your new kernels in the config file, after "_partial_ : True". Hydra tries to load the desired kernel in "models/encoder_processor_decoder.py". If the desired library isnt available, torch currently will fall back to torch.nn.<kernel>.

```python
        for kernel in self.layer_kernels:
            kernel_entry=self.layer_kernels[kernel]
            try:
                instantiate(kernel_entry)
            except InstantiationException:
                LOGGER.info(f"{kernel_entry['_target_']} not availible! falling back to torch.nn.{kernel}")
                #config.model.layer_kernels[kernel]["_target_"]=f"torch.nn.{kernel}"
                self.layer_kernels[kernel] = DotDict({'_target_': f"torch.nn.{kernel}", '_partial_': True}) #replace the entry, to remove any args passed to the orginal kernel
        LOGGER.debug(f"{self.layer_kernels=}")
```
I am in two minds about this. On the one hand, it helps ensure you run if you are missing a library (which might be nessecary when doing inference in a different enviroment to where the model was trained). But on the other hand, maybe it would be better to loudly fail when the user is requesting an uninstalled library be used. Otherwise, a user could be under the impression they are using an optimised kernel when they are not. Also it feels like poor programming practice to branch on Exceptions like I am doing here. Open to suggestions on this.

This feature makes it easy to try out new kernels in an end-to-end machine learning run, rather then simply doing a standalone kernel benchmark. Using this feature I was able to trial [the RMSNorm implementation of LayerNorm from Liger Kernel](https://github.com/linkedin/Liger-Kernel), in 2 lines (`pip install liger_kernel; vim config/model/transformer.yaml`) and I saw ~10% speedup. 

| LayerNorm | run_training_batch time (s) | training_avg_throughput (iter/s) |
| ------------- | ------------- | ------------- |
| torch.nn.LayerNorm |   0.93532 |0.88574 |
| liger_kernel.transformers.rms_norm.LigerRMSNorm | 0.82982 | 0.99186 |

# Type of change

Please delete options that are not relevant.

New feature (non-breaking change which adds functionality)

This change requires a documentation update

# Checklist before requesting a review

- [X] I have performed a self-review of my code
- [X] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation and docstrings to reflect the changes
- [ ]  have added tests that prove my fix is effective or that my feature works
- [X] I have ensured that the code is still pip-installable after the changes and runs
- [X] I have not introduced new dependencies in the inference partion of the model
- [X] I have ran this on single GPU
- [X] I have ran this on multi-GPU or multi-node
- [ ] I have ran this to work on LUMI (or made sure the changes work independently.)
- [X] I have ran the Benchmark Profiler against the old version of the code

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=35) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-models?pullRequest=35) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Cathal Liam O Brien<br/>:x: cathalobrien<br/><hr/>**Cathal Liam O Brien** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-models?pullRequest=35) it.</sub>

**Comment by clessig** [comment]
I think silently switching to a different implementation is dangerous. As we see with the attention, there might be differences in the implementation that the user should be aware of and think about when selecting a different backend.

**Comment by cathalobrien** [comment]
```python
LOGGER.info(f"{kernel_entry['_target_']} not availible! falling back to torch.nn.{kernel}")
```
I agree, good analogy with attention. At the moment there's a warning, but stdout is easily missed. 

However, we have to make sure inference is still possible without any additional libraries that might have been used during training. This could be done by resetting the 'layer_kernels' config entry to "torch.nn" during inference.  but at the moment there's apparently no easy way to tell if you're in inference or training, so this requires some thought

File: src/anemoi/models/layers/block.py [hunk 1]
```diff
@@ -316,15 +330,20 @@ def __init__(
...
         self.num_chunks = num_chunks
 
-        self.lin_key = nn.Linear(in_channels, num_heads * self.out_channels_conv)
-        self.lin_query = nn.Linear(in_channels, num_heads * self.out_channels_conv)
-        self.lin_value = nn.Linear(in_channels, num_heads * self.out_channels_conv)
-        self.lin_self = nn.Linear(in_channels, num_heads * self.out_channels_conv, bias=bias)
-        self.lin_edge = nn.Linear(edge_dim, num_heads * self.out_channels_conv)  # , bias=False)
+        # Uses the implementation defined in config.model.layer_kernels.<kernel>
+        # (unless it is not availible, in which case it will fall back to torch.nn.<kernel>)
+        Linear = layer_kernels["Linear"]
```
Comment (mishooax) [inline src/anemoi/models/layers/block.py:335]: are these keys guaranteed to be in the `layer_kernels` dict?

Comment (mishooax) [reply src/anemoi/models/layers/block.py:335]: possible alternative:

```python3
Linear = layer_kernels.get("Linear", torch.nn.Linear)
```

File: src/anemoi/models/layers/mlp.py [hunk 2]
```diff
@@ -64,23 +67,30 @@ def __init__(
...
+            mlp1.append(Linear(hidden_dim, hidden_dim))
             mlp1.append(act_func())
-        mlp1.append(nn.Linear(hidden_dim, out_features))
+        mlp1.append(Linear(hidden_dim, out_features))
 
         if final_activation:
             mlp1.append(act_func())
 
         if layer_norm:
             mlp1.append(AutocastLayerNorm(out_features))
```
Comment (mishooax) [inline src/anemoi/models/layers/mlp.py:92]: do we still want the `AutocastLayerNorm` here? should this be replaced by a `LayerNorm`?

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 3]
```diff
@@ -69,6 +70,31 @@ def __init__(
...
+            kernel_entry = self.layer_kernels[kernel]
+            try:
+                instantiate(kernel_entry)
+            except InstantiationException:
+                LOGGER.info(f"{kernel_entry['_target_']} not availible! falling back to torch.nn.{kernel}")
+                # config.model.layer_kernels[kernel]["_target_"]=f"torch.nn.{kernel}"
+                self.layer_kernels[kernel] = DotDict(
+                    {"_target_": f"torch.nn.{kernel}", "_partial_": True}
+                )  # replace the entry, to remove any args passed to the orginal kernel
+        LOGGER.debug(f"{self.layer_kernels=}")
```
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:96]: should be safe to remove this

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 4]
```diff
@@ -69,6 +70,31 @@ def __init__(
 
         self.num_channels = config.model.num_channels
 
+        # If self.layer_kernels entry is missing from the config, use torch.nn by default
```
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:73]: can you factor this out to a separate member function? the body of the init is rather long now

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 5]
```diff
@@ -69,6 +70,31 @@ def __init__(
...
+            self.layer_kernels = config.model.layer_kernels
+        except AttributeError:
+            LOGGER.info("No entry found for config.model.layer_kernels. Defaulting to torch.nn implementations")
+            self.layer_kernels = DotDict()
+            self.layer_kernels["Linear"] = DotDict({"_target_": "torch.nn.Linear", "_partial_": True})
+            self.layer_kernels["LayerNorm"] = DotDict({"_target_": "torch.nn.LayerNorm", "_partial_": True})
+
+        # try loading each of the requested kernels
+        # If a given kernel isnt availible, fallback to the torch.NN implementation of the same name
+        # TODO I would prefer to come up with a way for hydra to loop over options to instiate, rather then having to catch errors like this
```
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:84]: spell check needed; also, is the TODO out of scope for this PR?

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 6]
```diff
@@ -69,6 +70,31 @@ def __init__(
...
+
+        for kernel in self.layer_kernels:
+            kernel_entry = self.layer_kernels[kernel]
+            try:
+                instantiate(kernel_entry)
+            except InstantiationException:
+                LOGGER.info(f"{kernel_entry['_target_']} not availible! falling back to torch.nn.{kernel}")
+                # config.model.layer_kernels[kernel]["_target_"]=f"torch.nn.{kernel}"
```
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:92]: remove this line

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 4 above)
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:73]: where is the config? i'm not seeing it in this PR

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 7]
```diff
@@ -69,6 +70,31 @@ def __init__(
 
         self.num_channels = config.model.num_channels
 
+        # If self.layer_kernels entry is missing from the config, use torch.nn by default
+        try:
+            self.layer_kernels = config.model.layer_kernels
+        except AttributeError:
+            LOGGER.info("No entry found for config.model.layer_kernels. Defaulting to torch.nn implementations")
```
Comment (mishooax) [inline src/anemoi/models/models/encoder_processor_decoder.py:77]: better to error out gracefully, perhaps?

File: src/anemoi/models/layers/attention.py [hunk 8]
```diff
@@ -41,6 +41,7 @@ def __init__(
         is_causal: bool = False,
         window_size: Optional[int] = None,
         dropout: float = 0.0,
+        layer_kernels: any = None,
```
[the remark this answers is not in the document]
Comment (cathalobrien) [reply src/anemoi/models/layers/attention.py:44]: laziness :D that was when I was hacking stuff together, I'll replace with something better now. thanks for catching

File: src/anemoi/models/layers/block.py (hunk 1 above)
Comment (cathalobrien) [reply src/anemoi/models/layers/block.py:335]: Oh that's nice! I just pushed a version which uses torch.nn if nothing is set. But I'll replace it with that

File: src/anemoi/models/layers/mlp.py (hunk 2 above)
Comment (cathalobrien) [reply src/anemoi/models/layers/mlp.py:92]: Is that where we landed from the Nvidia meeting way back? I can do some tests

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 5 above)
Comment (cathalobrien) [reply src/anemoi/models/models/encoder_processor_decoder.py:84]: sure, It was more of a note to check if Hydra could do this. But I'll remove it

**Comment by cathalobrien** [comment]
> ```python
> LOGGER.info(f"{kernel_entry['_target_']} not availible! falling back to torch.nn.{kernel}")
> ```
> 
> I agree, good analogy with attention. At the moment there's a warning, but stdout is easily missed.
> 
> However, we have to make sure inference is still possible without any additional libraries that might have been used during training. This could be done by resetting the 'layer_kernels' config entry to "torch.nn" during inference. but at the moment there's apparently no easy way to tell if you're in inference or training, so this requires some thought

I'll make a PR to ai_models, where i reset layer_kernels to torch.nn if thay've been set. seems like the most straightforward way to handle inference

**Comment by JesperDramsch** [comment]
Open to a discussion, but I think we shouldn't have "default fallbacks" for these types of things, if the code fails to instantiate from the provided config.

When someone sbatch submits their job with a specific experiment, it should probably be obvious that the config wasn't valid, but with these fallbacks, it just "runs through" anyways and you have to monitor your logs if anything went wrong with your experiment due to a misconfiguration or wrong environment.

File: src/anemoi/models/layers/block.py (hunk 1 above)
Comment (cathalobrien) [reply src/anemoi/models/layers/block.py]: done

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 3 above)
Comment (cathalobrien) [reply src/anemoi/models/models/encoder_processor_decoder.py]: done

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 4 above)
Comment (cathalobrien) [reply src/anemoi/models/models/encoder_processor_decoder.py]: done, _load_layer_kernels()

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 7 above)
Comment (cathalobrien) [reply src/anemoi/models/models/encoder_processor_decoder.py]: good idea, done

**Comment by mchantry** [comment]
Now being implemented in #95
